### DISCUSSION TITLE
Can the Server side initiate a Stream and send data over it?

### DISCUSSION THREAD

**Discussion by chemsallioua** [microsoft/msquic#4257] [Q&A] [answered]
Hi, I am new to the MSQUIC library, and I am struggling to do just basic Server - Client connection.

I want the Server to listen for connections, and the Client to connect and open a QUIC Connection. When the connection is established, It's the server that should start STREAMS in the connection, and send data over it. I tried to do so, but on the client side i receive an Connection EVENT of type QUIC_CONNECTION_EVENT_PEER_NEEDS_STREAMS.

Does that mean that the Server cannot start streams, but it can express the need for streams and its the Cleint who actually has to open streams?

Thank you!

**Comment by nibanks** [comment]
Yep! The client just needs to configure its settings to allow for the server to create streams. Streams can always be created by either side, if the peer allows.

**Reply by anrossi** [reply, marked as answer]
Expanding on what Nick said, see [Settings](https://github.com/microsoft/msquic/blob/main/docs/Settings.md) and [QUIC_SETTINGS](https://github.com/microsoft/msquic/blob/main/docs/api/QUIC_SETTINGS.md).  The `PeerBidiStreamCount` and `PeerUniDiStreamCount` need to be non-zero on the client side to allow the server to open streams.

**Reply by chemsallioua** [reply]
Thank you for the very fast reply! That was what I needed.
Out of curiosity is there an app based on MSQUIC which facilitates the development of apps on top of them?
I am writing one starting from the sample available on the repo, but It is not tested and not complete.

I would appreciate your help.
